cheatsheets/Forgot_Password_Cheat_Sheet.md

```diff
@@ -1,60 +1,70 @@
 
-## Step 3) Send a Token Over a Side-Channel
+### One Time Password
 
-After step 2, lock out the user's account immediately. Then SMS or utilize some other multi-factor token challenge with a randomly-generated code having 8 or more characters. 
+OTP is the best method in order to implement a secure forgot password service that triggers as a 2FA functionality.
 
-This introduces an *out-of-band* communication channel and adds defense-in-depth as it is another barrier for a hacker to overcome. If the bad guy has somehow managed to successfully get past steps 1 and 2, he is unlikely to have compromised the side-channel. It is also a good idea to have the random code which your system generates to only have a limited validity period, say no more than 20 minutes or so. That way if the user doesn't get around to checking their email and their email account is later compromised, the random token used to reset the password would no longer be valid if the user never reset their password and the *reset password* token was discovered by an attacker. 
+The two most famous methods are Time-OTP ([TOTP](https://tools.ietf.org/html/rfc6238)), or HMAC-OTP ([HOTP](https://tools.ietf.org/html/rfc4226)). The main difference is in the counter, where TOTP focuses on the Unix time, and HOTP has a counter that gets incremented on every user call to generate the OTP.
 
-The user should always be sent a token rather than a new password for their account. There are numerous problems with directly sending the user a password, especially via email. These include:
+TOTP is a favorite candidate over HOTP, since you don't have to worry about the counter, and it gets refreshed between 30 or 60 seconds in most implementations. One implementation can be found over for [Authy](https://www.twilio.com/docs/authy/tutorials/two-factor-authentication-python-flask). If you don't want to rely on applications (such as Authy, Google/Microsoft Authenticator, etc.), you can generate QR codes and send them from the server for the user to scan using their device. The following python library, [pyotp](https://github.com/pyauth/pyotp), helps the developer implement any of the methods discussed in this section.
 
-- The transport mechanism used to send the password (email, SMS, etc) may not be encrypted.
-- The password may be stored in the users' mailbox or device for an indefinite period.
-- The password may be accessible to IT staff or other users who have access to the mailbox or device.
-- The password may be logged by third parties such as mail security scanning platforms.
+> OTPs can be sent through other channels as well, such as emails and SMSs. Various attacks and weaknesses have been identified in such technologies that it is preferrable not to use them for OTPs.
 
-Of course, by all means, once a user's password has been reset, the randomly-generated token should no longer be valid.
+### URL Tokens
```

Comment:
Also, I've seen some times using JWT for password reset. It provides a nice stateless way to have expiration dates without a backend DB, but has all the usual risks of JWT.